tools/testinfra/runner/src/main.rs

```diff
@@ -73,45 +88,79 @@ fn main() -> Result<()> {
     // don't run anything when just listing
     if options.list {
         for test in tests {
-            println!("{}", test.name);
+            println!("{}", test_name(&test.target, &test.unit));
         }
         exit(0);
     }
 
+    // connect to DB when it is provided
+    let mut db = match options.conn {
+        None => None,
+        Some(ref uri) => Some(
+            Client::connect(&uri, NoTls)
+                .with_context(|| format!("Couldn't connect to specified test DB at '{}'", uri))?,
+        ),
+    };
+    let disabled = query_disabled(&mut db);
+
     // run tests in parallel (retries share the same thread)
     let total = tests.len();
-    eprintln!("Running {} tests...", total);
+    eprintln!("Found {} tests...", total);
     let _ = rayon::ThreadPoolBuilder::new()
         .num_threads(options.threads)
         .build_global();
     let mut tests: Vec<TestResult> = tests
         .into_par_iter()
         .map(|test| {
-            let test = test.run(options.retries);
+            let run = !disabled.contains(&(test.target.clone(), test.unit.clone()))
+                || options.run_disabled;
+            let test = if run {
+                test.run(options.retries)
+            } else {
+                TestResult {
+                    target: test.target,
+                    unit: test.unit,
+                    attempts: 0,
+                    passed: false,
+                    duration: Duration::ZERO,
+                    stdout: "".to_string(),
+                    stderr: "".to_string(),
+                    contacts: test.contacts,
+                }
+            };
+
+            let name = test_name(&test.target, &test.unit);
             if test.passed {
-                print!("[OK] {} ({} ms)", test.name, test.duration.as_millis());
+                print!("[PASS] {} ({} ms)", name, test.duration.as_millis());
                 if test.attempts > 1 {
                     print!(" ({} attempts needed)\n", test.attempts);
                 } else {
                     print!("\n");
                 }
+            } else if test.attempts == 0 {
+                println!("[SKIP] {}", name);
             } else {
-                println!("[FAIL] {} ({} ms)", test.name, test.duration.as_millis());
+                println!("[FAIL] {} ({} ms)", name, test.duration.as_millis());
             }
+
             return test;
         })
         .collect();
 
     // collect and print results
     let mut passed = 0;
+    let mut skipped = 0;
     let mut errors: Vec<String> = Vec::new();
     for test in tests.iter_mut() {
```

Comment:
This doesn't seem like it has to be `iter_mut`. Also a `fold` with a tuple struct might be easier than incrementing many different variables.
You can use a `(usize, usize, Vec<String>)` tuple and `fold` on `tests.iter()` to mutate that tuple on each iteration based on the status